### PR TITLE
feat: x509 sdjwt and openid support

### PR DESCRIPTION
**Pull request by auer-martin**

### CODE REVIEW

File: packages/core/src/modules/sd-jwt-vc/SdJwtVcService.ts [hunk 1]
```diff
@@ -393,12 +423,28 @@ export class SdJwtVcService {
       }
     }
 
-    throw new SdJwtVcError("Unsupported credential issuer. Only 'did' is supported at the moment.")
+    if (issuer.method === 'x5c') {
+      const leafCertificate = X509Service.getLeafCertificate(agentContext, { certificateChain: issuer.chain })
+      const key = leafCertificate.publicKey
+      const alg = getJwkFromKey(key).supportedSignatureAlgorithms[0]
```
[the remark this answers is not in the document]
Comment (auer-martin) [reply packages/core/src/modules/sd-jwt-vc/SdJwtVcService.ts:429]: Not sure, which PR you mean

File: packages/core/src/modules/sd-jwt-vc/SdJwtVcOptions.ts [hunk 2]
```diff
@@ -30,12 +30,31 @@ export interface SdJwtVcIssuerDid {
...
 
+export interface SdJwtVcIssuerX5c {
+  method: 'x5c'
+  /**
+   *
+   * Array of base64-encoded certificate strings in the DER-format.
+   *
+   * The certificate containing the public key corresponding to the key used to digitally sign the JWS MUST be the first certificate.
+   */
+  chain: string[]
```
Comment (TimoGlastra) [inline packages/core/src/modules/sd-jwt-vc/SdJwtVcOptions.ts:41]: Should we also update this to x5c like we did in sphereon's libs?

Comment (auer-martin) [reply packages/core/src/modules/sd-jwt-vc/SdJwtVcOptions.ts:41]: yes I already did that locally. Pushing the changes as soon as the sphereon vp pr is merged.

File: packages/openid4vc/src/openid4vc-verifier/OpenId4VcSiopVerifierService.ts [hunk 3]
```diff
@@ -93,9 +93,60 @@ export class OpenId4VcSiopVerifierService {
...
+        const extractDnsNameFromDnsUriScheme = (iss: string): string => {
+          const afterDns = iss.substring('dns:'.length)
+          const withoutQuery = afterDns.split('?')[0]
+          return withoutQuery.split('/').pop() || ''
+        }
+
+        const dnsName = extractDnsNameFromDnsUriScheme(issuer)
+
+        // TODO: HAIP: x.509 certificates: the SD-JWT VC contains the issuer's certificate along with a trust chain in the x5c JOSE header.
+        // In this case, the iss value MUST be an URL with a FQDN matching a dNSName Subject Alternative Name (SAN) [RFC5280] entry in the leaf certificate.
```
Comment (TimoGlastra) [inline packages/openid4vc/src/openid4vc-verifier/OpenId4VcSiopVerifierService.ts:119]: Is this still todo?

File: packages/openid4vc/src/openid4vc-verifier/OpenId4VcSiopVerifierService.ts [hunk 4]
```diff
@@ -424,20 +462,21 @@ export class OpenId4VcSiopVerifierService {
       .withRedirectUri(authorizationResponseUrl)
       .withIssuer(ResponseIss.SELF_ISSUED_V2)
       .withSupportedVersions([SupportedVersion.SIOPv2_D11, SupportedVersion.SIOPv2_D12_OID4VP_D18])
-      .withCustomResolver(getSphereonDidResolver(agentContext))
-      .withResponseMode(ResponseMode.POST)
+      .withResponseMode(ResponseMode.DIRECT_POST)
```
Comment (TimoGlastra) [inline packages/openid4vc/src/openid4vc-verifier/OpenId4VcSiopVerifierService.ts:465]: Is this a breaking change?

Comment (TimoGlastra) [reply packages/openid4vc/src/openid4vc-verifier/OpenId4VcSiopVerifierService.ts:465]: Will it still work with older versions (let's say we update Paradym Wallet first and Paradym later, or vice versa?)

File: packages/openid4vc/src/openid4vc-verifier/__tests__/openid4vc-verifier.test.ts [hunk 5]
```diff
@@ -65,7 +65,7 @@ describe('OpenId4VcVerifier', () => {
       expect(jwt.payload.additionalClaims.redirect_uri).toEqual(
         `http://redirect-uri/${openIdVerifier.verifierId}/authorize`
       )
-      expect(jwt.payload.additionalClaims.response_mode).toEqual('post')
+      expect(jwt.payload.additionalClaims.response_mode).toEqual('direct_post')
```
Comment (TimoGlastra) [inline packages/openid4vc/src/openid4vc-verifier/__tests__/openid4vc-verifier.test.ts:68]: this feels like a breaking change. Do older versions of the oidp4v-sip lib already support direct_post?

File: packages/openid4vc/src/shared/models/OpenId4VcJwtIssuer.ts [hunk 6]
```diff
@@ -1,13 +1,33 @@
...
+   *
+   * The certificate containing the public key corresponding to the key used to digitally sign the JWS MUST be the first certificate.
+   */
+  x5c: string[]
+
+  /**
+   * The issuer of the JWT. Should be a HTTPS URI.
+   *
+   * The issuer value must either match a `uniformResourceIdentifier` SAN entry of the leaf entity certificate
+   * or match the domain name in the `iss` value matches the `dNSName` SAN entry of the end-entity certificate.
```
Comment (TimoGlastra) [inline packages/openid4vc/src/shared/models/OpenId4VcJwtIssuer.ts:23]:
```suggestion
   * or the domain name in the `iss` value matches the `dNSName` SAN entry of the end-entity certificate.
```

File: packages/openid4vc/src/shared/utils.ts [hunk 7]
```diff
@@ -34,61 +38,95 @@ export function getSupportedJwaSignatureAlgorithms(agentContext: AgentContext):
...
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    const { iss, sub, aud, exp, nbf, iat, jti, ...additionalClaims } = jwt.payload
+
+    const jwsService = agentContext.dependencyManager.resolve(JwsService)
+
+    if (jwtIssuer.method === 'did') {
+      const key = await getKeyFromDid(agentContext, jwtIssuer.didUrl)
+      const jws = await jwsService.createJwsCompact(agentContext, {
+        protectedHeaderOptions: { alg: jwtIssuer.alg, ...jwt.header },
+        payload: new JwtPayload({ ...jwt.payload, additionalClaims }),
```
Comment (TimoGlastra) [inline packages/openid4vc/src/shared/utils.ts:78]: You can use JwtPayload.fromJson here? so you don't have to spread all the properties

File: packages/openid4vc/src/shared/utils.ts [hunk 8]
```diff
@@ -34,61 +38,95 @@ export function getSupportedJwaSignatureAlgorithms(agentContext: AgentContext):
...
+
+    return {
+      method: openId4VcTokenIssuer.method,
+      didUrl: openId4VcTokenIssuer.didUrl,
+      alg: _alg as unknown as SigningAlgo,
+    }
+  } else if (openId4VcTokenIssuer.method === 'x5c') {
+    return {
+      ...openId4VcTokenIssuer,
+      clientIdScheme: 'x509_san_dns',
```
Comment (TimoGlastra) [inline packages/openid4vc/src/shared/utils.ts:125]: In other code I saw you also handled `x509_san_uri`. Shouldn't we handle that here as well?

**Review by TimoGlastra — COMMENTED** [review]
Good stuff, and the new APIs in sphereon's lib work really well here!! :)

File: packages/core/src/modules/sd-jwt-vc/SdJwtVcOptions.ts (hunk 2 above)
Comment (TimoGlastra) [inline packages/core/src/modules/sd-jwt-vc/SdJwtVcOptions.ts:41]: This is still `chain` ( i see you updated it in the verifier part)

File: packages/openid4vc/src/openid4vc-verifier/OpenId4VcSiopVerifierService.ts [hunk 9]
```diff
@@ -424,20 +462,21 @@ export class OpenId4VcSiopVerifierService {
       .withRedirectUri(authorizationResponseUrl)
       .withIssuer(ResponseIss.SELF_ISSUED_V2)
       .withSupportedVersions([SupportedVersion.SIOPv2_D11, SupportedVersion.SIOPv2_D12_OID4VP_D18])
```
Comment (TimoGlastra) [inline packages/openid4vc/src/openid4vc-verifier/OpenId4VcSiopVerifierService.ts:464]: We also support Draft 20 now right? We should add it here then

File: packages/openid4vc/src/openid4vc-verifier/__tests__/openid4vc-verifier.test.ts (hunk 5 above)
Comment (auer-martin) [reply packages/openid4vc/src/openid4vc-verifier/__tests__/openid4vc-verifier.test.ts:68]: support is there starting from v17

File: packages/openid4vc/src/openid4vc-verifier/OpenId4VcSiopVerifierService.ts (hunk 3 above)
Comment (auer-martin) [reply packages/openid4vc/src/openid4vc-verifier/OpenId4VcSiopVerifierService.ts:119]: yes. with haip it cannot be a san uri name

File: packages/core/src/modules/x509/X509ModuleConfig.ts [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    this.options.trustedCertificates = trustedCertificates ? [...trustedCertificates] : undefined
+  }
+
+  public addTrustedCertificate(trustedCertificate: string) {
+    if (!this.options.trustedCertificates) {
+      this.options.trustedCertificates = [trustedCertificate]
+      return
+    }
+    this.options.trustedCertificates.push(trustedCertificate)
+  }
```
Comment (TimoGlastra) [inline packages/core/src/modules/x509/X509ModuleConfig.ts:26]: You are still overwriting the input options. This is not immutable. We should add an `#trustedCertificates` parameter to this class that is initlaized based on the [...options.trustedCertificates] . And when you call set/add you overwrite the / push to #trustedCertificates. This way the user input is not modified

File: packages/core/src/modules/x509/X509Service.ts [hunk 11]
```diff
@@ -50,12 +51,29 @@ export class X509Service {
...
     // The chain is reversed here as the `x5c` header (the expected input),
     // has the leaf certificate as the first entry, while the `x509` library expects this as the last
-    const parsedChain = chain.map((c) => X509Certificate.fromRawCertificate(new Uint8Array(c.rawData))).reverse()
+    let parsedChain = chain.map((c) => X509Certificate.fromRawCertificate(new Uint8Array(c.rawData))).reverse()
 
     if (parsedChain.length !== certificateChain.length) {
       throw new X509Error('Could not parse the full chain. Likely due to incorrect ordering')
     }
 
+    if (trustedCertificates) {
```
[the remark this answers is not in the document]
Comment (TimoGlastra) [reply packages/core/src/modules/x509/X509Service.ts:60]: This doesn't seem solved, or I can't see the solution

File: packages/openid4vc/src/openid4vc-verifier/OpenId4VcSiopVerifierService.ts [hunk 12]
```diff
@@ -93,9 +93,60 @@ export class OpenId4VcSiopVerifierService {
...
+    let clientId: string
+    let clientIdScheme: ClientIdScheme
+
+    if (requestSigner.method === 'x5c') {
+      const issuer = requestSigner.issuer
+      const leafCertificate = X509Service.getLeafCertificate(agentContext, {
+        certificateChain: requestSigner.x5c,
+      })
+
+      if (issuer.startsWith('dns:')) {
```
[the remark this answers is not in the document]
Comment (TimoGlastra) [reply packages/openid4vc/src/openid4vc-verifier/OpenId4VcSiopVerifierService.ts]: Is this now solved by requiring the user to pass the clientIdScheme? Do we not check anymore the x5c certiifcate against  the issuer?

[the remark this answers is not in the document]
Comment (auer-martin) [reply packages/openid4vc/src/openid4vc-verifier/OpenId4VcSiopVerifierService.ts]: no? where have you seen that?

File: packages/core/src/modules/x509/X509Service.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (auer-martin) [reply packages/core/src/modules/x509/X509Service.ts:60]: I think this logic should not be on the x509 service. It is clear IMO that it cannot perform the validation there if no trusted certificates are passed. I check in the sdjwtvc service for it.

```ts
      const trustedCertificates = agentContext.dependencyManager.resolve(X509ModuleConfig).trustedCertificates
      if (!trustedCertificates) {
        throw new SdJwtVcError(
          'No trusted certificates configured for X509 certificate chain validation. Issuer cannot be verified.'
        )
      }
```

File: packages/openid4vc/src/openid4vc-verifier/OpenId4VcSiopVerifierService.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (TimoGlastra) [reply packages/openid4vc/src/openid4vc-verifier/OpenId4VcSiopVerifierService.ts]: ah wait seems you extracted it to a separate method 👍

File: packages/core/src/modules/x509/X509Service.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (TimoGlastra) [reply packages/core/src/modules/x509/X509Service.ts:60]: Hm okay. I'm not sure. I think not disabling validation should be explicity. But let's keep it like this for now. The validation is done at least 👍

**Review by TimoGlastra — APPROVED** [review]

**Pull request merged**